### PR TITLE
2. change(state): Run AwaitUtxo read requests without shared mutable chain state

### LINKED ISSUE #5102: Make StateService requests into concurrent ReadStateService requests
## Motivation

At the moment, Zebra reads the state via two different services:
- a blocking `StateService`
- a concurrent `ReadStateService`

Most read requests are able to be run concurrently, so we could move them to the `ReadStateService`. This allows us to move mutable chain state into the block commit task.

This should also improve performance.

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

This PR runs `StateService::AwaitUtxo` read requests without accessing shared mutable chain state.
As a useful side-effect, this makes all `StateService` read requests concurrent.

Closes #5102.

### Designs

See #4937 and https://docs.google.com/drawings/d/1FXpAUlenDAjl8nkftrypdAPsj0jr-Ut9gZlSP57nuyc/edit

## Solution

Don't access shared mutable chain state in `StateService` read requests:
- Re-Implement `AwaitUtxo` without accessing shared mutable chain state, using `StateService.pending_utxos` and `ReadStateService::AnyUtxo`
  - Send the whole `NonFinalizedState` to the `ReadStateService`
  - Rename the `ChainUtxo` request to `BestChainUtxo`
  - Add an `AnyChainUtxo` request

Related documentation:
- Update the state service module documentation
- Explain how spent UTXOs are treated by the state
- Clarify how cached Chains impact state read requests

Related Cleanups:
- Move `AwaitUtxos` next to the other shared writeable state requests
- Rename `ReadResponse::Utxos` to `ReadResponse::AddressUtxos`
- Rename an `out_point` variable to `outpoint` for consistency
- Rename `transparent_utxos` to `address_utxos`
- Move the `QueuedBlock` type into the `queued_blocks` module

## Testing

We should run a full sync on this before we merge, and make sure it is not significantly slower than the `main` branch.

## Review

This PR is part of regular scheduled work.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Actually create a separate task to commit blocks to the state.

### CODE REVIEW

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
I'm running a full sync and a checkpoint sync here:
https://github.com/ZcashFoundation/zebra/actions/runs/3013393462

**Comment by teor2345** [comment]
Weird, this built on my machine. I'll push my latest changes and see what's up.

Edit: it was some newly created files I forgot to commit.

**Comment by teor2345** [comment]
I'm running a full sync and a checkpoint sync here:
https://github.com/ZcashFoundation/zebra/actions/runs/3018034862

**Comment by teor2345** [comment]
I did a full sync and an update sync locally:
- both succeeded on testnet
- update sync succeeded on mainnet
- full sync is still going on mainnet, it seems to be about the same speed, there are no warnings

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs]:
```suggestion
    ///
    /// TODO: move users of read [`Request`]s to [`ReadStateService`], and remove `read_service`.
```

File: zebra-state/src/service/read/find.rs
Comment (teor2345) [inline zebra-state/src/service/read/find.rs:78]: Explain why doing this concurrently is ok.

File: zebra-state/src/service/read/find.rs
Comment (teor2345) [inline zebra-state/src/service/read/find.rs:162]: Explain why doing this concurrently is ok.
Is it ok to access hashes by height if the finalized state is changing?

File: zebra-state/src/request.rs
Comment (teor2345) [inline zebra-state/src/request.rs]:
```suggestion
    /// Checks the finalized chain, all non-finalized chains, queued unverified blocks,
```

File: zebra-state/src/response.rs
Comment (teor2345) [inline zebra-state/src/response.rs]:
```suggestion
    /// The response to a `AwaitUtxo` request, from any non-finalized chains, finalized chain,
```

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs]: Work out how to access all the chains without using `self.mem`

File: zebra-state/src/service.rs
Comment (teor2345) [inline zebra-state/src/service.rs:712]: Open a ticket for this?

File: zebra-state/src/service/read/find.rs
Comment (teor2345) [reply zebra-state/src/service/read/find.rs:78]: Added a comment in 47011e1a52

File: zebra-state/src/service/read/find.rs
Comment (teor2345) [reply zebra-state/src/service/read/find.rs:162]: Added a comment in 47011e1a52

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs:712]: I opened #5126.

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs]: Fixed in ecb0259246 by sending the whole `NonFinalizedState` to the `ReadStateService`.

**Review by arya2 — DISMISSED** [review]
The logic in `Tip`, `Depth`, `BlockLocator`, `FindBlockHashes`, and `FindBlockHeaders` requests of `ReadStateService` looks to be the same, and the only difference I see in the `AwaitUtxo` request of `StateService` is that it iterates through `queued_blocks` before checking `any_utxo()`, but that shouldn't cause issues.

The `TryFrom` impls for `Request` and `ReadResponse` look good.

The new dead code and the unused `ChainUtxo` request in `ReadStateService` are better than potentially forgetting we had those later and re-writing similar logic.

File: zebra-state/src/service.rs
Comment (arya2) [inline zebra-state/src/service.rs:689]: it would be nice if some of these metrics counters were moved up to cut down on repetition

File: zebra-state/src/service.rs
Comment (arya2) [inline zebra-state/src/service.rs]: would be nice to delete the `metric::counter!()` calls in the arm below to shorten the code a bit here too

```suggestion
    fn call(&mut self, req: ReadRequest) -> Self::Future {
        metrics::counter!(
            "state.requests",
            1,
            "service" => "read_state",
            "type" => match req {
                ReadRequest::Depth(_) => "depth",
                ReadRequest::Tip => "tip",
                ReadRequest::BlockLocator => "block_locator",
                ReadRequest::Transaction(_) => "transaction",
                ReadRequest::ChainUtxo { .. } => "chain_utxo",
                ReadRequest::Block(_) => "block",
                ReadRequest::FindBlockHashes { .. } => "find_block_hashes",
                ReadRequest::FindBlockHeaders { .. } => "find_block_headers",
                ReadRequest::SaplingTree { .. } => "sapling_tree",
                ReadRequest::OrchardTree { .. } => "orchard_tree",
                ReadRequest::AddressBalance { .. } => "address_balance",
                ReadRequest::TransactionIdsByAddresses { .. } => "transaction_ids_by_addesses",
                ReadRequest::UtxosByAddresses(_) => "utxos_by_addesses"
            }
        );
```

File: zebra-state/src/service.rs
Comment (arya2) [inline zebra-state/src/service.rs]: Not sure if the Requests that are now redirected to the ReadStateService should be counted as both state service and read_state, but might be nicer to have it in a `.count_metric()` method than a `.name()` method

```suggestion
        metrics::counter!(
            "state.requests",
            1,
            "service" => "state",
            "type" => match req {
                Request::Depth(_) => "depth",
                Request::Tip => "tip",
                Request::BlockLocator => "block_locator",
                Request::Transaction(_) => "transaction",
                Request::Block(_) => "block",
                Request::FindBlockHashes { .. } => "find_block_hashes",
                Request::FindBlockHeaders { .. } => "find_block_headers",
                Request::CommitBlock(_) => "commit_block",
                Request::CommitFinalizedBlock(_) => "commit_finalized_block",
                Request::AwaitUtxo(_) => "await_utxo",
            }
        );

```

File: zebra-state/src/service.rs
Comment (arya2) [inline zebra-state/src/service.rs]: Following the suggestion to add the metrics::counter at the top of `.call()`

```suggestion
            Request::Depth(_)
            | Request::Tip
            | Request::BlockLocator
            | Request::Transaction(_)
            | Request::Block(_)
            | Request::FindBlockHashes { .. }
            | Request::FindBlockHeaders { .. } => {
                // Redirect the request to the concurrent ReadStateService
                let read_service = self.read_service.clone();
                async move {
                    let req = req
                        .try_into()
                        .expect("ReadRequest conversion should not fail");
                    let rsp = read_service.oneshot(req).await?;
                    let rsp = rsp.try_into().expect("Response conversion should not fail");
                    Ok(rsp)
                }
                .boxed()
            }

```

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs]: Would you mind pushing a separate PR with these changes?
They are a good cleanup, but they don't block the rest of the work in #4937, so we can merge them separately.

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs]: These changes can also go in that separate PR.

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs]: These changes can also go in that separate PR.

**Comment by teor2345** [comment]
The previous full sync was about 10% slower during full verification.

I'm running a full sync again with the latest code:
https://github.com/ZcashFoundation/zebra/actions/runs/3025949651

I have some ideas to fix that, but I think they should go in a separate PR.
We might also see a significant speedup once we implement a separate task for block commits.

So I might need to temporarily increase the full sync timeout in this PR, just to make `main` pass while we're doing these other changes.

File: zebra-state/src/service/read/block.rs
Comment (upbqdn) [inline zebra-state/src/service/read/block.rs]: Consistency note: other comments say "are".
```suggestion
/// UTXOs are returned regardless of whether they have been spent.
```

File: zebra-state/src/service/read/block.rs
Comment (upbqdn) [inline zebra-state/src/service/read/block.rs]: Consistency note: other comments say "are".
```suggestion
/// UTXOs are returned regardless of whether they have been spent.
```

File: zebra-state/src/service/non_finalized_state.rs
Comment (upbqdn) [inline zebra-state/src/service/non_finalized_state.rs:349]: The doc comment for this function says it returns
> the [`transparent::Utxo`] pointed to by the given [`transparent::OutPoint`]

However, it returns such a UTXO from the first chain when there are multiple parallel non-finalized chains containing the same UTXO.

I checked whether this affects anything, and it doesn't seem so.

**Comment by teor2345** [comment]
> The previous full sync was about 10% slower during full verification.

Actually, it seems like full syncs have been failing on `main` for a while now, so I pushed some PRs to fix that.

File: zebra-state/src/service/read/block.rs
Comment (upbqdn) [inline zebra-state/src/service/read/block.rs]: I'm probably missing something, but it's not clear to me what `latest chain` refers to.

File: zebra-state/src/service/read/find.rs
Comment (upbqdn) [inline zebra-state/src/service/read/find.rs:105]: I'm probably missing something, but it's not clear to me what `overlapping blocks` refers to.

File: zebra-state/src/service/read/find.rs
Comment (upbqdn) [inline zebra-state/src/service/read/find.rs:125]: I'm probably missing something, but it's not clear to me what `overlapping blocks` refers to.

File: zebra-state/src/service/read/find.rs
Comment (upbqdn) [inline zebra-state/src/service/read/find.rs:91]: I'm probably missing something, but it's not clear to me what `overlapping blocks` refers to.

File: zebra-state/src/service/read/find.rs
Comment (upbqdn) [inline zebra-state/src/service/read/find.rs:76]: I'm probably missing something, but it's not clear to me what `overlapping blocks` refers to.

File: zebra-state/src/service/read/find.rs
Comment (upbqdn) [inline zebra-state/src/service/read/find.rs:152]: I'm probably missing something, but it's not clear to me what `overlapping blocks` refers to.

**Comment by teor2345** [comment]
I split PR #5132 out of this PR, because it seems like we've finished work on most of the state requests, so they can merge. But we're still working on the `AwaitUtxos` request.

**Review by upbqdn — COMMENTED** [review]
Looks great so far. I need a bit more time to finish reviewing the `AwaitUtxo` requests.

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs]: > Not sure if the Requests that are now redirected to the ReadStateService should be counted as both state service and read_state

I think it's ok for now, because some requests (`Block` and `Transaction`) are used via both `StateService` and `ReadStateService` directly. So their metrics will be different across the services.

Eventually I'd like to use `ReadStateService` directly, get rid of the read requests in the `StateService`, and then the corresponding metrics will go as well. But that's something for a future cleanup.

File: zebra-state/src/service/non_finalized_state.rs
Comment (teor2345) [reply zebra-state/src/service/non_finalized_state.rs:349]: The `OutPoint` contains the transaction hash, which includes all the content of the `Utxo`. So there is only ever one `Utxo` per `OutPoint`, even across multiple chains.

File: zebra-state/src/service/read/block.rs
Comment (teor2345) [reply zebra-state/src/service/read/block.rs]: This was actually a deliberate difference, but you're right - it's not clear what it means.

I explained what actually happens in commit 1d4668432a

File: zebra-state/src/service/read/block.rs
Comment (teor2345) [reply zebra-state/src/service/read/block.rs]: I explained what actually happens in commit 1d4668432a

File: zebra-state/src/service/read/block.rs
Comment (teor2345) [reply zebra-state/src/service/read/block.rs]: It's the cached `Chain` in the `chain` variable passed to this function.
Hopefully the doc changes in commit 8015f51c48 make this a bit clearer.

File: zebra-state/src/service/read/find.rs
Comment (teor2345) [reply zebra-state/src/service/read/find.rs:76]: If the request is delayed, the cached `chain` can contain some of the same blocks as the shared `db` reference.
Hopefully the doc changes in commit 8015f51c48 make this a bit clearer.

File: zebra-state/src/service/read/find.rs
Comment (teor2345) [reply zebra-state/src/service/read/find.rs:91]: See commit 8015f51c48.

File: zebra-state/src/service/read/find.rs
Comment (teor2345) [reply zebra-state/src/service/read/find.rs:105]: See commit 8015f51c48.

File: zebra-state/src/service/read/find.rs
Comment (teor2345) [reply zebra-state/src/service/read/find.rs:125]: See commit 8015f51c48.

File: zebra-state/src/service/read/find.rs
Comment (teor2345) [reply zebra-state/src/service/read/find.rs:152]: See commit 8015f51c48.

**Comment by teor2345** [comment]
I made `AwaitUtxo` non-blocking, that should be enough to get us started with block commit channels.

File: zebra-state/src/service/non_finalized_state.rs
Comment (upbqdn) [reply zebra-state/src/service/non_finalized_state.rs:349]: I meant the scenario when there is the same transaction in two parallel chains, for example when two separate miners produce two separate blocks with some transactions being the same. Does your claim hold in this scenario as well?

Comment (upbqdn) [reply zebra-state/src/service/non_finalized_state.rs:349]: But as I said, I don't think this scenario is important because it doesn't affect Zebra's behavior. I just wanted to understand what actually happens.

**Review by arya2 — DISMISSED** [review]
Looks good, I added some cleanup in [minor cleanup of zebra-state/service.rs](https://github.com/ZcashFoundation/zebra/pull/5137) for later

File: zebra-state/src/service/non_finalized_state.rs
Comment (teor2345) [reply zebra-state/src/service/non_finalized_state.rs:349]: > I meant the scenario when there is the same transaction in two parallel chains, for example when two separate miners produce two separate blocks with some transactions being the same. Does your claim hold in this scenario as well?

In this case:
- if the transactions are the same, the `OutPoint`s are the same, and they return the same `Utxo`
- if the transactions are different, the transaction hash or index in the `OutPoint`s are different, even if the `Utxo`s have the same value and spend script

https://github.com/ZcashFoundation/zebra/blob/97fb85dca9d00a717661ea0cbab7ce66b8d23977/zebra-chain/src/transparent.rs#L99-L113

**Comment by teor2345** [comment]
> Run gcloud compute instances create-with-container "update-to-tip-5107-merge-7ae4b28" \
> ERROR: (gcloud.compute.instances.create-with-container) Could not fetch resource:
>  - The resource 'projects/zealous-zebra/global/images/zebrad-cache-5143-merge-fb82d20-v25-mainnet-tip-u-013716' is not ready

https://github.com/ZcashFoundation/zebra/actions/runs/3041636484/jobs/4899242654

This will be fixed by #5044

**Comment by teor2345** [comment]
@mergifyio update

File: zebra-state/src/service/non_finalized_state.rs
Comment (upbqdn) [reply zebra-state/src/service/non_finalized_state.rs:349]: Got it, thank you for taking the time to explain it.

**Comment by teor2345** [comment]
I'm running a full sync here, but we don't need to wait for it before we merge this PR:
https://github.com/ZcashFoundation/zebra/actions/runs/3047965824

File: zebra-state/src/service/read/block.rs
Comment (upbqdn) [inline zebra-state/src/service/read/block.rs:6]: I still have trouble with `latest chains`.

Does it refer to the most recently updated chain in the `chain_set` in the `NonFinalizedState`?

File: zebra-state/src/service.rs
Comment (upbqdn) [inline zebra-state/src/service.rs:167]: Maybe we could remove `recent` since there are no other `NonFinalizedState`s?

It made me think if there are any non-recent `NonFinalizedState`s.

File: zebra-state/src/service.rs
Comment (upbqdn) [inline zebra-state/src/service.rs:703]: I wonder if this actually returns `Some` in practice.

The reason is that I'd expect almost all UTXOs to be at least a few blocks apart.

**Review by upbqdn — APPROVED** [review]
All looks good. I left some minor comments.

File: zebra-state/src/service/read/block.rs
Comment (teor2345) [reply zebra-state/src/service/read/block.rs:6]: I'll make this change to all these files in my editor.
At the same time, I'll change `StateService` to "block commit task", because that's what it will be after this PR series.

```suggestion
//! `chain` or `non_finalized_state` with a cached copy of the non-finalized chains in `NonFinalizedState.chain_set`. Then the block commit task can
```

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs:167]:
```suggestion
    /// A watch channel with a cached copy of the [`NonFinalizedState`].
```

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs:703]: Here's one way it could return `Some`:
1. The syncer sends 20 blocks for full verification in parallel
2. The first block takes a while to verify
3. The fifth block gets verified quickly, because it doesn't depend on anything from the first four blocks, then it gets put in the queue, waiting for the first four blocks
4. The tenth block spends a UTXO from the fifth block, which is in the queue, so `AwaitUtxo` returns `Some` here

Feel free to add a debug log and see how often that happens!

File: zebra-state/src/service.rs
Comment (teor2345) [reply zebra-state/src/service.rs:167]: Updated in PR #5134 in commit 22ef6561f8

File: zebra-state/src/service/read/block.rs
Comment (teor2345) [reply zebra-state/src/service/read/block.rs:6]: Updated in PR #5134 in commit 22ef6561f8

File: zebra-state/src/service.rs
Comment (upbqdn) [reply zebra-state/src/service.rs:703]: Makes sense!

**Pull request merged**
